File: windows/ProLicense.cs
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;
namespace TokFire.Bench;
internal sealed class ProLicense {
 readonly JsonObject cfg=JsonNode.Parse(File.ReadAllText(Path.Combine(AppContext.BaseDirectory,"runner","lemon-squeezy.json")))!.AsObject();
 string FilePath=>Path.Combine(AppSettings.Root,"license.dat");
 public bool Configured=>new[]{"storeId","productId","variantId"}.All(k=>(cfg[k]?.GetValue<int>()??0)>0);
 public string? Credential {get;private set;}
 public ProLicense(){try{Credential=Encoding.UTF8.GetString(ProtectedData.Unprotect(File.ReadAllBytes(FilePath),null,DataProtectionScope.CurrentUser));}catch{Credential=null;}}
 public async Task Activate(string key){
  if(!Configured)throw new InvalidOperationException("Live Pro sales are not yet configured.");
  if(!Guid.TryParse(key,out _))throw new InvalidOperationException("Enter a valid license key.");
  var r=await Call("activate",new(){["license_key"]=key,["instance_name"]="TokFire Windows "+Guid.NewGuid().ToString("N")[..8]});
  var instance=r["instance"]?["id"]?.GetValue<string>();
  if(r["activated"]?.GetValue<bool>()!=true||r["error"] is not null||instance is null)throw new InvalidOperationException("Activation failed. Check the key and one-device activation limit.");
  var credential=new JsonObject{["key"]=key,["instanceID"]=instance}.ToJsonString();
  try{await Validate(credential);}catch{await Call("deactivate",new(){["license_key"]=key,["instance_id"]=instance});throw;}
  Credential=credential;File.WriteAllBytes(FilePath,ProtectedData.Protect(Encoding.UTF8.GetBytes(credential),null,DataProtectionScope.CurrentUser));
 }
 public async Task Validate(string? value=null){
  if(!Configured||string.IsNullOrEmpty(value??Credential))throw new InvalidOperationException("Activate Pro to select more than three jobs.");
  var c=JsonNode.Parse(value??Credential!)!;var r=await Call("validate",new(){["license_key"]=c["key"]!.GetValue<string>(),["instance_id"]=c["instanceID"]!.GetValue<string>()});
  if(r["valid"]?.GetValue<bool>()!=true||r["error"] is not null||r["license_key"]?["status"]?.GetValue<string>()!="active"||r["license_key"] is not JsonObject license||!license.ContainsKey("expires_at")||license["expires_at"] is not null||r["instance"]?["id"]?.GetValue<string>()!=c["instanceID"]!.GetValue<string>()||r["license_key"]?["key"]?.GetValue<string>()!=c["key"]!.GetValue<string>())throw new InvalidOperationException("Pro license is inactive or invalid.");
  foreach(var pair in new[]{("store_id","storeId"),("product_id","productId"),("variant_id","variantId")})if(r["meta"]?[pair.Item1]?.GetValue<int>()!=cfg[pair.Item2]!.GetValue<int>())throw new InvalidOperationException("License is for another product.");
 }
 public async Task Deactivate(){if(Credential is null)return;var c=JsonNode.Parse(Credential)!;var r=await Call("deactivate",new(){["license_key"]=c["key"]!.GetValue<string>(),["instance_id"]=c["instanceID"]!.GetValue<string>()});if(r["deactivated"]?.GetValue<bool>()!=true)throw new InvalidOperationException("Deactivation failed; the key remains saved.");File.Delete(FilePath);Credential=null;}
 static async Task<JsonObject> Call(string action,Dictionary<string,string> data){using var client=new HttpClient{Timeout=TimeSpan.FromSeconds(25)};using var response=await client.PostAsync("https://api.lemonsqueezy.com/v1/licenses/"+action,new FormUrlEncodedContent(data));response.EnsureSuccessStatusCode();return JsonNode.Parse(await response.Content.ReadAsStringAsync())!.AsObject();}
}

File: native/Sources/LocalAIBench/Resources/pro_license.py
"""Lemon Squeezy License API. No seller API token or customer identity is stored."""
import json, urllib.request, urllib.parse, urllib.error, uuid
from pathlib import Path

def config():
    return json.loads(Path(__file__).with_name('lemon-squeezy.json').read_text())

def validate_response(value, cfg, credential):
    license=value.get('license_key',{});meta=value.get('meta',{});instance=value.get('instance') or {}
    if (value.get('valid') is not True or value.get('error') is not None
        or license.get('status') != 'active' or license.get('key') != credential['key']
        or 'expires_at' not in license or license['expires_at'] is not None
        or instance.get('id') != credential['instanceID']):
        raise ValueError('An active perpetual Pro license instance is required.')
    for field,setting in [('store_id','storeId'),('product_id','productId'),('variant_id','variantId')]:
        expected=cfg.get(setting)
        if type(expected) is not int or expected<=0 or type(meta.get(field)) is not int or meta[field]!=expected:
            raise ValueError('This license is not for TokFire Bench Pro.')
    return True

def verify_license(serialized):
    cfg=config()
    if any(type(cfg.get(k)) is not int or cfg[k]<=0 for k in ['storeId','productId','variantId']):
        raise ValueError('Lemon Squeezy Pro product is not configured in this build.')
    credential=json.loads(serialized)
    try: uuid.UUID(credential['key']);uuid.UUID(credential['instanceID'])
    except (KeyError,ValueError,TypeError):raise ValueError('Invalid Pro activation.') from None
    body=urllib.parse.urlencode({'license_key':credential['key'],'instance_id':credential['instanceID']}).encode()
    req=urllib.request.Request('https://api.lemonsqueezy.com/v1/licenses/validate',body,headers={'Content-Type':'application/x-www-form-urlencoded','Accept':'application/json'})
    try:
        with urllib.request.urlopen(req,timeout=20) as response:
            raw=response.read(65537)
            if len(raw)>65536:raise ValueError('Invalid license response.')
            return validate_response(json.loads(raw),cfg,credential)
    except (urllib.error.URLError,TimeoutError):
        raise ValueError('Unable to verify Pro with Lemon Squeezy. Check your activation and internet connection.') from None

def authorize(count, key=None):
    if type(count) is not int or not 1<=count<=20:raise ValueError('Select 1–20 concurrent jobs.')
    if count>3:
        if not key:raise ValueError('More than three jobs requires a Lemon Squeezy Pro license.')
        verify_license(key)
